pkg/doc/verifiable/credential.go

```diff
@@ -242,83 +242,36 @@ type rawCredential struct {
 	Issuer         interface{}       `json:"issuer,omitempty"`
 	Schema         interface{}       `json:"credentialSchema,omitempty"`
 	Evidence       *Evidence         `json:"evidence,omitempty"`
-	TermsOfUse     []TermsOfUse      `json:"termsOfUse,omitempty"`
-	RefreshService *RefreshService   `json:"refreshService,omitempty"`
+	TermsOfUse     []TypedID         `json:"termsOfUse,omitempty"`
+	RefreshService *TypedID          `json:"refreshService,omitempty"`
 
 	// All unmapped fields are put here.
 	CredentialExtraFields `json:"-"`
 }
 
-func (rc *rawCredential) marshalJSON() ([]byte, error) {
-	// Convert raw credential into a JSON map of known fields.
-	nf, err := rc.asMap()
-	if err != nil {
-		return nil, err
-	}
+// MarshalJSON defines custom marshalling of rawCredential to JSON.
+func (rc *rawCredential) MarshalJSON() ([]byte, error) {
+	type Alias rawCredential
+	alias := (*Alias)(rc)
 
-	// Supplement raw credential map with unknown fields.
-	uf := rc.CredentialExtraFields
-	for k, v := range uf {
-		if _, exists := nf[k]; !exists {
-			nf[k] = v
-		}
-	}
-
-	// Marshal extended known fields map.
-	return json.Marshal(nf)
+	return marshalWithExtraFields(alias, rc.CredentialExtraFields)
 }
 
-func (rc *rawCredential) asMap() (map[string]interface{}, error) {
-	// Convert raw credential into a JSON map of known fields.
-	rcBytes, err := json.Marshal(rc)
-	if err != nil {
-		return nil, err
-	}
+// UnmarshalJSON defines custom unmarshalling of rawCredential from JSON.
+func (rc *rawCredential) UnmarshalJSON(data []byte) error {
+	type Alias rawCredential
+	alias := (*Alias)(rc)
 
-	var nf map[string]interface{}
-	err = json.Unmarshal(rcBytes, &nf)
+	ef := make(CredentialExtraFields)
+	err := unmarshalWithExtraFields(data, alias, ef)
 	if err != nil {
-		return nil, err
-	}
-
-	return nf, nil
-}
-
-func newRawCredential(bytes []byte) (*rawCredential, error) {
-	rc := new(rawCredential)
-	err := json.Unmarshal(bytes, rc)
-	if err != nil {
-		return nil, fmt.Errorf("raw verifiable credential: %w", err)
-	}
-
-	// Collect known fields map.
-	rcBytes, err := json.Marshal(rc)
-	if err != nil {
-		return nil, err
-	}
-
-	var kf map[string]interface{}
-	err = json.Unmarshal(rcBytes, &kf)
-	if err != nil {
-		return nil, err
-	}
-
-	// Collect all fields map.
-	var af map[string]interface{}
-	err = json.Unmarshal(bytes, &af)
-	if err != nil {
-		return nil, err
-	}
-
-	// Complement all fields with known fields to become a set of unknown fields.
-	for k := range kf {
-		delete(af, k)
+		return err
 	}
 
-	// Save unknown fields.
-	rc.CredentialExtraFields = af
+	*rc = *(*rawCredential)(alias)
```

Comment:
same here, Unmarshal changes the (pointer) value of the main struct *rawCredential ..

Reply:
fixed